Set optimal shifts

GuyDrukman opened the pull request.

GuyDrukman committed:
Added Dockerfile + Added setShifts algorithm

GuyDrukman committed:
Fix Typo

GuyDrukman committed:
Fix Typo

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +75,65 @@ public void deleteShift(RoutingContext ctx) {
+            ShiftWeightSettings settings = shiftWeightSettingsService.getSettings();
+            String currentPreset = settings.getCurrentPreset();
+            List<ShiftWeight> shiftsWeight= settings.getPresets().get(currentPreset).getWeights();
+            List<User> users = userService.getAllUsers();
```
I will send a List of User - those are users that we need to assign in this week.
We iterate on that list we will get from the ConstraintsService.getUserConstraints() the constraints for each user.
This will be a Map<User, Constraint[]>

GuyDrukman committed:
Improved algorithm

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +78,135 @@ public void deleteShift(RoutingContext ctx) {
+            ShiftWeightSettings settings = objectMapper.readValue(body, ShiftWeightSettings.class);
+            List<Constraint> constraints = objectMapper.readValue(body, objectMapper.getTypeFactory().constructCollectionType(List.class, Constraint.class));
+
+            constraintService.addConstraints(constraints);
```
Why?

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +78,135 @@ public void deleteShift(RoutingContext ctx) {
+                }
+            }
+
+            shiftService.addShifts(assignedShifts);
```
Don't do that, the list of assigned shifts will be sent to the client as a suggestion, the user will decide of he saves that or not

GuyDrukman committed:
Improved algorithm

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +78,138 @@ public void deleteShift(RoutingContext ctx) {
+                    logger.warn("No suitable user found for shift: " + newShift.getDate() + " " + newShift.getType());
+                }
+            }
+
```
That too should only happen when the user submits the shifts, need to add in the shiftService.addShift()

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +78,138 @@ public void deleteShift(RoutingContext ctx) {
+
+            for (User user : users) {
+                JsonObject updates = new JsonObject().put("score", user.getScore());
+                userService.updateUser(user.getId(), updates);
```
Send the array of assignedShifts here

henashke commented on src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java:
```diff
@@ -68,10 +78,138 @@ public void deleteShift(RoutingContext ctx) {
+                userMissedDays.put(user.getId(), 0);
+            }
+
+            for (ShiftWeight shiftWeight : shiftsWeight) {
```
Here I can send you startDate, endDate from the client and you would generate the list of shifts to assign from that

henashke committed:
Refactor `ShiftHandler` for shift assignment suggestions and improve `UserService` ID management

- Renamed `setShifts` to `suggestShiftAssignment` in `ShiftHandler` and updated related routing.
- Refactored shift assignment logic for improved clarity and modularity.
- Replaced `User.id` usage with `User.name` as the unique identifier across services.
- Removed unused `Konan`-related fields and methods in `User`.

GuyDrukman committed:
Improved algorithm

GuyDrukman committed:
Merge remote-tracking branch 'origin/Set-Optimal-Shifts' into Set-Optimal-Shifts

# Conflicts:
#	src/main/java/com/shiftmanagerserver/handlers/ShiftHandler.java
